### PR TITLE
Issue 175, no BCL references for .Net40

### PR DESCRIPTION
**Pull request by ki0ki0**
BCL dependencies has been removed from base API project.
TelemetryChannel still uses BCL.

[Issue 175](https://github.com/Microsoft/ApplicationInsights-dotnet/issues/175)

### CODE REVIEW

**Comment by abaranch** [comment]
Thank you for your contribution. We will review it soon.

File: Test/CoreSDK.Test/Net40/TaskExTests.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    using Assert = Xunit.Assert;
+
+    [TestClass]
+    public class TaskExTests
+    {
+
+        [TestMethod]
+        public void RethrowIfFailedDoesntThrowIfNoExceptionOccured()
+        {
+            var task = Task.Factory.StartNew(() => { });
```
Comment (dnduffy) [inline Test/CoreSDK.Test/Net40/TaskExTests.cs]: nitpick: I know this is a hotly debated topic online but I respectfully ask that you use concrete types instead of var as it increases readability of the code (especially outside of Visual Studio where there is no IntelliSense).

File: Test/CoreSDK.Test/Net40/TaskExTests.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
+namespace Microsoft.ApplicationInsights
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+    using Assert = Xunit.Assert;
+
+    [TestClass]
```
Comment (dnduffy) [inline Test/CoreSDK.Test/Net40/TaskExTests.cs]: Documentation comments and general StyleCop compliance would be really appreciated.  Thank-you. :)

File: Test/CoreSDK.Test/Net40/TaskExTests.cs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        [TestMethod]
+        public void DelayWhichIsCanceledThrows()
+        {
+            var timeout = TimeSpan.FromSeconds(1);
+            var cancellationTokenSource = new CancellationTokenSource();
+            var task = TaskEx.Delay(timeout, cancellationTokenSource.Token);
+
+            cancellationTokenSource.Cancel();
+
+            Assert.Throws<AggregateException>(() => task.Wait());
```
Comment (dnduffy) [inline Test/CoreSDK.Test/Net40/TaskExTests.cs]: To be safe, the test should look into the AggregateException and make sure the cancellation is indeed the reason.

File: Test/CoreSDK.Test/Net40/TaskExTests.cs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+            {
+                task.Wait();
+            }
+            catch (Exception)
+            {
+                // ignore
+            }
+            var spent = DateTime.UtcNow - time;
+
+            Assert.True(spent >= cancelTimeout && spent < timeout);
```
Comment (dnduffy) [inline Test/CoreSDK.Test/Net40/TaskExTests.cs]: It's probably safe enough for the test but just FYI this is technically a form of race condition that could fail in an edge case situation of high CPU load.

File: Test/CoreSDK.Test/Net40/TaskExTests.cs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+            task.Wait();
+
+            Assert.DoesNotThrow(() => task.RethrowIfFailed());
+        }
+
+        [TestMethod]
+        public void RethrowIfFailedThrowsIfTaskIsNotCompleted()
+        {
+            var task = TaskEx.Delay(TimeSpan.FromMilliseconds(100));
```
Comment (dnduffy) [inline Test/CoreSDK.Test/Net40/TaskExTests.cs]: In later files you have replaced TaskEx with Task so why use TaskEx here?

File: Test/CoreSDK.Test/Net40/TaskExTests.cs (hunk 1 above)
Comment (macrogreg) [inline Test/CoreSDK.Test/Net40/TaskExTests.cs]: Usually "var" is OK as long as the type is clear from the *same* line without specific knowledge of the code (and in complex Linq expressions).
var x = new MyType(); // this is OK
var y = MyFunc(); // Not OK since it requires knowing what MyFunc returns.
this particular one is on the border since Task.Factory.StartNew is "almost" like new task :)

Comment (SergeyKanzhelev) [inline Test/CoreSDK.Test/Net40/TaskExTests.cs]: I agree with @macrogreg - we are using `var` widely in the code already.

File: src/Core/Managed/Net40/TaskEx.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            Timer timer = null;
+            timer = new Timer(
+                state =>
+                    {
+                        timer.Dispose();
+                        tcs.TrySetResult(null);
+                    }, 
+                null, 
+                timeout, 
+                TimeSpan.FromMilliseconds(-1));
```
Comment (macrogreg) [inline src/Core/Managed/Net40/TaskEx.cs]: Please use "Timeout.Infinite"

File: src/Core/Managed/Net40/TaskEx.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        public static Task<Task> WhenAny(params Task[] tasks)
+        {
+            if (tasks.Length == 0)
+            {
+                throw new ArgumentException("The tasks argument contains no tasks");
+            }
+
+            var taskCompletionSource = new TaskCompletionSource<Task>();
+
+            foreach (var task in tasks)
```
Comment (macrogreg) [inline src/Core/Managed/Net40/TaskEx.cs]: Micro-optimization: use a for loop to avoid an enumerator object creation.

File: src/Core/Managed/Shared/Channel/InMemoryTransmitter.cs [hunk 8]
```diff
@@ -133,18 +133,22 @@ private void DequeueAndSend(TimeSpan timeout)
...
+#else
+                return Task.FromResult<object>(null);
+#endif
             }
 
             byte[] data = JsonSerializer.Serialize(telemetryItems);
             var transmission = new Transmission(this.endpointAddress, data, "application/x-json-stream", JsonSerializer.CompressionType, timeout);
 
-            await transmission.SendAsync().ConfigureAwait(false);
+            return transmission.SendAsync();
```
Comment (macrogreg) [inline src/Core/Managed/Shared/Channel/InMemoryTransmitter.cs]: This is behavior change: the original await cause any exceptions that occurred in the async part of  SendAsync to be thrown here. Now they will instead be thrown when the Task returned by SendAsync is awaited. Can this be a problem?

File: src/Core/Managed/Shared/Channel/Transmission.cs [hunk 9]
```diff
@@ -148,27 +153,39 @@ public string Id
...
-                                
-                Task completedTask = await TaskEx.WhenAny(timeoutTask, sendTask).ConfigureAwait(false);
-                if (completedTask == timeoutTask)
+                Task timeoutTask = TaskEx.Delay(this.Timeout).ContinueWith(task =>
                 {
-                    request.Abort(); // And force the sendTask to throw WebException.
-                }
+                    if (!sendTask.IsCompleted)
+                    {
+                        request.Abort(); // And force the sendTask to throw WebException.
```
Comment (macrogreg) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: There is a race condition here on sendTask completing after the check. I do not recall what happens if Abort is called on a finished request. If nothing, this is benign. To avoid this you can use a synchronous continuation to the send task and null out the request reference or set an interlocked flag. And yes - the issue existed before the change. :)

File: src/Core/Managed/Shared/Channel/Transmission.cs [hunk 10]
```diff
@@ -131,14 +131,19 @@ public string Id
...
+#else
         public virtual async Task<HttpWebResponseWrapper> SendAsync()
+#endif
         {
             if (Interlocked.CompareExchange(ref this.isSending, 1, 0) != 0)
             {
                 throw new InvalidOperationException("SendAsync is already in progress.");
             }
-
+#if !NET40
```
Comment (macrogreg) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: due to the complex #if , this method is hard to read and understand. I would prefer to see a global (method-level) #if and then two clean implementations for 4.0 and not 4.0.

File: src/Core/Managed/Shared/Channel/Transmission.cs [hunk 11]
```diff
@@ -148,27 +153,39 @@ public string Id
...
-                    request.Abort(); // And force the sendTask to throw WebException.
-                }
+                    if (!sendTask.IsCompleted)
+                    {
+                        request.Abort(); // And force the sendTask to throw WebException.
+                    }
+                });
+
+#if NET40
+                return TaskEx.WhenAny(timeoutTask, sendTask).ContinueWith(
```
Comment (macrogreg) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: Use synchronous continuation to unset the flag.

File: src/Core/Managed/Shared/Channel/Transmission.cs [hunk 12]
```diff
@@ -221,6 +238,38 @@ protected virtual WebRequest CreateRequest(Uri address)
...
         }
 
+#if NET40
+        private Task<HttpWebResponseWrapper> SendRequestAsync(WebRequest request)
+        {
+            return Task.Factory.FromAsync(request.BeginGetRequestStream, request.EndGetRequestStream, null)
+                .ContinueWith(
+                    task =>
+                    {
+                        var requestStream = task.Result;
```
Comment (macrogreg) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: Here, the type is not clear from the same line. Please do not use VAR.

File: src/Core/Managed/Shared/Channel/Transmission.cs [hunk 13]
```diff
@@ -221,6 +238,38 @@ protected virtual WebRequest CreateRequest(Uri address)
...
+                        return Task.Factory.FromAsync(
+                            (callback, o) => requestStream.BeginWrite(Content, 0, Content.Length, callback, o),
+                            requestStream.EndWrite, 
+                            null).ContinueWith(
+                                task1 =>
```
Comment (dnduffy) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: To avoid confusion I would rename task1 to something else such as innerTask.

File: src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs [hunk 14]
```diff
@@ -77,7 +81,33 @@ public void Start(Func<Task> elapsed)
                             // It should return Task.FromResult but just in case we check for null if someone returned null
                             if (task != null)
                             {
+#if !NET40
                                 await task.ConfigureAwait(false);
+#else
+                                task.ContinueWith(
+                                    task1 =>
```
Comment (dnduffy) [inline src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs]: Please rename task1 to avoid confusion for future readers.

File: Test/CoreSDK.Test/Net40/TaskExTests.cs (hunk 1 above)
Comment (dnduffy) [inline Test/CoreSDK.Test/Net40/TaskExTests.cs]: Like I said, I know it is a hotly debated topic.  I am however very firmly on the side of readability and the framework design guidelines that Microsoft used to publish online and later in book form.

File: src/Core/Managed/Shared/Channel/Transmission.cs [hunk 15]
```diff
@@ -221,6 +238,38 @@ protected virtual WebRequest CreateRequest(Uri address)
             return request;
         }
 
+#if NET40
+        private Task<HttpWebResponseWrapper> SendRequestAsync(WebRequest request)
+        {
+            return Task.Factory.FromAsync(request.BeginGetRequestStream, request.EndGetRequestStream, null)
+                .ContinueWith(
+                    task =>
```
Comment (macrogreg) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: can we call this "getRequestStreamTask"?

File: src/Core/Managed/Shared/Channel/Transmission.cs (hunk 13 above)
Comment (macrogreg) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: a better name would be "writeTask"

File: src/Core/Managed/Net40/TaskEx.cs [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    internal static class TaskEx
+    {
+        public static void RethrowIfFailed(this Task task)
+        {
+            if (!task.IsCompleted)
```
Comment (macrogreg) [inline src/Core/Managed/Net40/TaskEx.cs]: The method is throw *ifFaied* so if the task is not competed, it has not failed and nothing should happen.

Comment (macrogreg) [inline src/Core/Managed/Net40/TaskEx.cs]: Also, if you want to emulate the behavior of await, yo need to unwrap the aggregate exception.

File: src/Core/Managed/Shared/Channel/Transmission.cs [hunk 17]
```diff
@@ -221,6 +238,38 @@ protected virtual WebRequest CreateRequest(Uri address)
...
+                            requestStream.EndWrite, 
+                            null).ContinueWith(
+                                task1 =>
+                                {
+                                    requestStream.Dispose();
+                                    task1.RethrowIfFailed();
+                                });
+                    }).Unwrap().ContinueWith(task =>
+                    {
+                        task.RethrowIfFailed();
```
Comment (macrogreg) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: You know that the task is completed. You could get the exception by asking for the result and unwrapping the aggregate (to be like await). Maybe adjust the implementation of RethrowIfFailed?

File: src/Core/Managed/Shared/Channel/Transmission.cs [hunk 18]
```diff
@@ -221,6 +238,38 @@ protected virtual WebRequest CreateRequest(Uri address)
...
+                        return Task.Factory.FromAsync(
+                            request.BeginGetResponse, request.EndGetResponse, null);
+                    }).Unwrap().ContinueWith(task =>
+                    {
+                        using (var response = task.Result)
```
Comment (macrogreg) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: What is the type of response? Please do not use VAR here

File: src/Core/Managed/Shared/Channel/Transmission.cs [hunk 19]
```diff
@@ -221,6 +238,38 @@ protected virtual WebRequest CreateRequest(Uri address)
...
+                        {
+                            return CheckResponse(response);
+                        }
+                    });
+        }
+#else
```
Comment (macrogreg) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: Please add a comment to say that "else" means "not 4.0"

File: src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs [hunk 20]
```diff
@@ -66,7 +66,11 @@ public void Start(Func<Task> elapsed)
 
             TaskEx.Delay(this.Delay, newTokenSource.Token)
                 .ContinueWith(
-                    async previousTask => 
+#if !NET40
+                async previousTask =>
+#else
+                    previousTask =>
```
Comment (macrogreg) [inline src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs]: Re line 78 below:
This is why VAR is so bad. I am reading that (unchanged) bit of code and I need to stop and think about the type returned by elapsed(). I have to look around and rely on correctly used conventions for a correct guess. :)

File: src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs [hunk 21]
```diff
@@ -77,7 +81,33 @@ public void Start(Func<Task> elapsed)
...
                             {
+#if !NET40
                                 await task.ConfigureAwait(false);
+#else
+                                task.ContinueWith(
+                                    task1 =>
+                                    {
+                                        try
+                                        {
+                                            task1.Wait();
```
Comment (macrogreg) [inline src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs]: you use Wait to hause exceptions to propagate. Elsewhere you use RethrowIf.... for that. Why the difference?

File: src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs [hunk 22]
```diff
@@ -66,7 +66,11 @@ public void Start(Func<Task> elapsed)
 
             TaskEx.Delay(this.Delay, newTokenSource.Token)
                 .ContinueWith(
-                    async previousTask => 
+#if !NET40
```
Comment (macrogreg) [inline src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs]: I have reservations about this entire method's existence, but the translation seems correct. :)

File: Test/CoreSDK.Test/Net40/TaskExTests.cs (hunk 2 above)
Comment (abaranch) [inline Test/CoreSDK.Test/Net40/TaskExTests.cs]: You could run build in release. Stylecop is automatically checked.

---
In reply to: [73396854](https://github.com/Microsoft/ApplicationInsights-dotnet/pull/267#discussion_r73396854) [](ancestors = 73396854)

File: src/Core/Managed/Shared/Channel/InMemoryTransmitter.cs (hunk 8 above)
Comment (abaranch) [inline src/Core/Managed/Shared/Channel/InMemoryTransmitter.cs]: We should not change the behavior.

---
In reply to: [73466495](https://github.com/Microsoft/ApplicationInsights-dotnet/pull/267#discussion_r73466495) [](ancestors = 73466495)

**Comment by SergeyKanzhelev** [comment]
@ki0ki0 - are you still interested in completing this PR? There are some comments that needs to be addressed before we can merge.

File: Test/CoreSDK.Test/Net40/TaskExTests.cs (hunk 1 above)
Comment (ki0ki0) [inline Test/CoreSDK.Test/Net40/TaskExTests.cs]: I've replaced all var where type is not clear from assignment

File: Test/CoreSDK.Test/Net40/TaskExTests.cs (hunk 2 above)
Comment (ki0ki0) [inline Test/CoreSDK.Test/Net40/TaskExTests.cs]: fixed

File: src/Core/Managed/Net40/TaskEx.cs (hunk 6 above)
Comment (ki0ki0) [inline src/Core/Managed/Net40/TaskEx.cs]: fixed

File: src/Core/Managed/Net40/TaskEx.cs (hunk 7 above)
Comment (ki0ki0) [inline src/Core/Managed/Net40/TaskEx.cs]: I've replaced implantation with ContinueWhenAny

File: src/Core/Managed/Shared/Channel/InMemoryTransmitter.cs (hunk 8 above)
Comment (ki0ki0) [inline src/Core/Managed/Shared/Channel/InMemoryTransmitter.cs]: The only usage of this private method is private method DequeueAndSend witch handles all exceptions.
So there are no changes in behavior.

File: src/Core/Managed/Shared/Channel/Transmission.cs (hunk 9 above)
Comment (ki0ki0) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: Abort on completed request doesn't have any effects,

File: src/Core/Managed/Shared/Channel/Transmission.cs (hunk 10 above)
Comment (ki0ki0) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: fixed

File: src/Core/Managed/Shared/Channel/Transmission.cs (hunk 11 above)
Comment (ki0ki0) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: fixed

File: src/Core/Managed/Shared/Channel/Transmission.cs (hunk 12 above)
Comment (ki0ki0) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: fixed

File: src/Core/Managed/Shared/Channel/Transmission.cs (hunk 13 above)
Comment (ki0ki0) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: fixed

File: src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs (hunk 14 above)
Comment (ki0ki0) [inline src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs]: fixed

File: src/Core/Managed/Shared/Channel/Transmission.cs (hunk 15 above)
Comment (ki0ki0) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: Renamed to GetResponseAsync. This method is implemented as async method for NET45 so I didn't remove Async suffix.

File: src/Core/Managed/Shared/Channel/Transmission.cs (hunk 17 above)
Comment (ki0ki0) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: Unwrapping of AggregateException for Net40 is tricky. I don't think it is a good idea.
IMHO RethrowIfFaulted is better option than checking result of task just to get an exception.

File: src/Core/Managed/Shared/Channel/Transmission.cs (hunk 18 above)
Comment (ki0ki0) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: fixed

File: src/Core/Managed/Shared/Channel/Transmission.cs (hunk 19 above)
Comment (ki0ki0) [inline src/Core/Managed/Shared/Channel/Transmission.cs]: fixed

File: src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs (hunk 20 above)
Comment (ki0ki0) [inline src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs]: fixed

**Comment by ki0ki0** [comment]
Sorry, I had no time to finish it.
I've fixed all issues and rebased on current "develop" position.
